components/brave_federated/learning_service.cc

```diff
@@ -0,0 +1,173 @@
+  DCHECK(participating_);
+  participating_ = false;
+}
+
+void LearningService::GetTasks() {
+  communication_adapter_->GetTasks(base::BindOnce(
+      &LearningService::HandleTasksOrReconnect, base::Unretained(this)));
+}
+
+void LearningService::HandleTasksOrReconnect(TaskList tasks, int reconnect) {
+  if (tasks.size() == 0) {
+    reconnect_timer_ = std::make_unique<base::RetainingOneShotTimer>();
+    reconnect_timer_->Start(FROM_HERE, base::Seconds(reconnect), this,
+                            &LearningService::GetTasks);
+    VLOG(2) << "No tasks available, reconnecting in " << reconnect << "s";
+    return;
+  }
+
+  // TODO(lminto): for now, disregard all tasks beyond the first one
+  // TODO(stevelaskaridis): extract into a utility function
+  Task task = tasks.at(0);
+  float lr = 0.01;
+  std::map<std::string, float> config = task.GetConfig();
+  auto cursor = config.find("lr");
+  if (cursor != config.end()) {
+    lr = cursor->second;
+    VLOG(2) << "Learning rate applied from server: " << lr;
+  }
+
+  ModelSpec spec{
+      32,    // num_params
+      64,    // batch_size
+      lr,    // learning_rate
+      500,   // num_iterations
+      0.5    // threshold
+  };
+
+  if (spec.num_params != static_cast<int>(task.GetParameters().at(0).size())) {
+    VLOG(2) << "Task specifies a different model size than the client";
+    return;
+  }
+  VLOG(2) << "Task model and client model match!";
+
+  Model* model = new Model(spec);
+  model->SetWeights(task.GetParameters().at(0));
+  model->SetBias(task.GetParameters().at(1).at(0));
+  FederatedTaskRunner* task_runner = new FederatedTaskRunner(task, model);
+
+  base::ThreadPool::PostTaskAndReplyWithResult(
+      FROM_HERE,
+      {base::TaskPriority::BEST_EFFORT,
+       base::TaskShutdownBehavior::CONTINUE_ON_SHUTDOWN},
+      base::BindOnce(
+          [](FederatedTaskRunner* task_runner) -> SyntheticDataset* {
+            return new SyntheticDataset(500);
+          }, std::move(task_runner)),
+      base::BindOnce(&LearningService::OnSyntheticDataGenerated,
+                     base::Unretained(this), std::move(task_runner)));
+}
+
+void LearningService::OnSyntheticDataGenerated(FederatedTaskRunner* task_runner, SyntheticDataset* local_dataset) {
+  SyntheticDataset local_test_data = local_dataset->SeparateTestData(50);
+
+  task_runner->SetTrainingData(local_dataset->GetDataPoints());
+  task_runner->SetTestData(local_test_data.GetDataPoints());
+  VLOG(2) << "Model and data set. Task runner initialized.";
+
+  base::ThreadPool::PostTaskAndReplyWithResult(
+      FROM_HERE,
+      {base::TaskPriority::BEST_EFFORT,
+       base::TaskShutdownBehavior::CONTINUE_ON_SHUTDOWN},
+      base::BindOnce(
+          [](FederatedTaskRunner* task_runner) -> TaskResult {
+            return task_runner->Run();
+          }, std::move(task_runner)),
+      base::BindOnce(&LearningService::OnTaskResultComputed,
+                     base::Unretained(this)));
```

Comment:
`base::Unretained` is also incorrect here.

Reply:
is this because this callback will be executed from another thread? `LearningService` itself is meant to be a persistent service (not destructed). What should we use here?

Reply:
What is the lifetime of `LearningService`? Is it a leaky singleton or something like this?
I thought it's bind to the profile.

Reply:
Yeah, it is bound to the profile